### ISSUE TITLE
Provide a way to reset S3 state before each test

### ISSUE THREAD

**Issue by vlsi** [adobe/S3Mock#229]
Use case: I want to run a parameterized test via JUnit5, so a single method runs multiple times.
Unfortunately, s3mock state is not reset across executions, so subsequent tests observe leftovers from the previous tests.

It would be nice to have something like `com.adobe.testing.s3mock.testsupport.common.S3MockStarter.BaseBuilder#withResetBeforeEachTest`

WDYT?

**Comment by vlsi** [comment]
It would be great if multiple tests that run concurrently could operate on their own isolated state.

For instance, what if there was a special connection `header` that specifies "virtual session id", so different clients could be isolated.

**Comment by afranken** [comment]
Thanks for the feature request.
Going forward, it is recommended not to start the S3Mock directly, but to start the S3Mock Docker container, for example through [TestContainers](https://www.testcontainers.org/) or by using a plugin in the build process such as the [Fabric8 docker-maven-plugin](https://github.com/fabric8io/docker-maven-plugin/).

The S3Mock is then accessible through its HTTP API, where only buckets and their objects are accessible.

**Comment by afranken** [comment]
Regarding your second comment:
Wouldn't it be sufficient to use different buckets?

**Comment by vlsi** [comment]
I'm not sure there's a way to use different buckets in parameterized test. Would you please clarify?

**Comment by vlsi** [comment]
The thing is `S3MockExtension.Builder` does not work as "test instance" extension since it starts mock server in `beforeAll`, so if I move `S3MockExtension.Builder` to the test instance, I get an NPE:

```
Caused by: java.lang.NullPointerException
	at com.adobe.testing.s3mock.testsupport.common.S3MockStarter.getPort(S3MockStarter.java:120)
	at com.adobe.testing.s3mock.testsupport.common.S3MockStarter.getServiceEndpoint(S3MockStarter.java:162)
	at com.adobe.testing.s3mock.testsupport.common.S3MockStarter.createS3ClientV2(S3MockStarter.java:82)
	at com.adobe.testing.s3mock.junit5.S3MockExtension.resolveParameter(S3MockExtension.java:114)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.resolveParameter(ExecutableInvoker.java:216)
```

```java
  public int getPort() {
    return s3MockFileStore.getPort(); // <-- NPE
  }
```

**Comment by afranken** [comment]
honestly, I don't understand the problem - in your test method, create the buckets you need, create the objects in those buckets. Use the test method name as the bucket name, for example.

Why would you need more than one S3Mock instance to test on multiple buckets?

**Comment by vlsi** [comment]
Here's the test class: https://github.com/burrunan/gradle-s3-build-cache/blob/224a366517bbb7321763e37cb380986bd9697345/src/test/kotlin/com/github/burrunan/s3cache/RemoteCacheTest.kt#L52-L61

I do not control what I put in the bucket. It is Gradle who decides what to store and retrieve from the bucket.
At the same time, different tests might happen to use the same objects within the bucket (Gradle uses content-addressed storage, so I can't easily configure it to use different object names).

So the only feasible option for me with the current code is:
a) Clean the bucket at the beginning of each test, and ensure the tests do not run in parallel
b) Use per-test buckets. However, currently, the S3 extension **must** be configured at the class level, so it would be inconvenient to specify "tmp bucket names when creating the extension, and when configuring the test".

Just in case you wonder, I have my own `@BeforeEach` callback that removes all the objects from the bucket (which is basically "a"): https://github.com/burrunan/gradle-s3-build-cache/blob/224a366517bbb7321763e37cb380986bd9697345/src/test/kotlin/com/github/burrunan/s3cache/RemoteCacheTest.kt#L93-L107
